### PR TITLE
refactor(tree):clean up flat tree example

### PR DESCRIPTION
**Pull request by vivian-hu**
clean up flat tree example and use element to update doc site

### CODE REVIEW

File: src/dev-app/tree/tree-demo-module.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ import {ChecklistTreeDemo} from './checklist-tree-demo/checklist-tree-demo';
 import {ChecklistNestedTreeDemo} from './checklist-tree-demo/checklist-nested-tree-demo';
 import {DynamicTreeDemo} from './dynamic-tree-demo/dynamic-tree-demo';
 import {LoadmoreTreeDemo} from './loadmore-tree-demo/loadmore-tree-demo';
-
+import {ExamplePageModule} from '../example/example-module';
```
Comment (jelbourn) [inline src/dev-app/tree/tree-demo-module.ts]: This should not be imported here. This file is auto-generated to import all of the individual example NgModules, so you're creating a circular dependency.

File: src/lib/tree/data-source/flat-data-source.ts [hunk 2]
```diff
@@ -50,7 +50,7 @@ export class MatTreeFlattener<T, F> {
   constructor(public transformFunction: (node: T, level: number) => F,
               public getLevel: (node: F) => number,
               public isExpandable: (node: F) => boolean,
-              public getChildren: (node: T) => Observable<T[]> | T[]) {}
+              public getChildren: (node: T) => Observable<T[]> | T[] | undefined) {}
```
Comment (jelbourn) [inline src/lib/tree/data-source/flat-data-source.ts]: The fix for this file should really be its own PR since it's a separate bug fix from cleaning up the examples. It also needs a unit test. I'd allow `null` here as well.

File: src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts [hunk 3]
```diff
@@ -1,113 +1,49 @@
...
-  },
-  Downloads: {
-    October: 'pdf',
-    November: 'pdf',
-    Tutorial: 'html'
+const TREE_DATA2: FoodNode[] = [
+  {
+    name: 'Fruit',
+    children: [
+      { name: 'Apple' },
```
Comment (jelbourn) [inline src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts]: Omit spaces in braces
(here and elsewhere)

File: src/material-examples/tree-flat-overview/tree-flat-overview-example.html [hunk 4]
```diff
@@ -1,16 +1,21 @@
...
     </button>
-    {{node.filename}} : {{node.type}}
+    {{node.name}}
   </mat-tree-node>
 </mat-tree>
+
+
+<!-- Copyright 2018 Google Inc. All Rights Reserved.
+    Use of this source code is governed by an MIT-style license that
+    can be found in the LICENSE file at http://angular.io/license -->
```
Comment (jelbourn) [inline src/material-examples/tree-flat-overview/tree-flat-overview-example.html]: The copyright is added automatically, you don't have to add it to the source.

File: src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts [hunk 5]
```diff
@@ -1,113 +1,49 @@
...
-        checkbox: 'ts',
-        input: 'ts'
-      }
-    }
-  },
-  Downloads: {
-    October: 'pdf',
-    November: 'pdf',
-    Tutorial: 'html'
+const TREE_DATA2: FoodNode[] = [
```
Comment (jelbourn) [inline src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts]: I would just call this `foodData`
(similar for other examples)

File: src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts [hunk 6]
```diff
@@ -117,33 +53,27 @@ export class FileDatabase {
...
-  constructor(database: FileDatabase) {
-    this.treeFlattener = new MatTreeFlattener(this.transformer, this._getLevel,
-      this._isExpandable, this._getChildren);
-    this.treeControl = new FlatTreeControl<FileFlatNode>(this._getLevel, this._isExpandable);
-    this.dataSource = new MatTreeFlatDataSource(this.treeControl, this.treeFlattener);
-
-    database.dataChange.subscribe(data => {
-      this.dataSource.data = data;
-    });
+  private _transformer = (node: FoodNode, level: number) => {
```
Comment (jelbourn) [inline src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts]: Don't use underscores for private members in the examples; that's just a convention for our libraries. 
(here and elsewhere)

Comment (jelbourn) [inline src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts]: Add a comment that explains what `transformer` does
(here and for similar code in other examples)

File: src/dev-app/tree/tree-demo.html [hunk 7]
```diff
@@ -1,22 +1,11 @@
...
-        <button mat-icon-button matTreeNodeToggle
-                [attr.aria-label]="'toggle ' + node.filename">
-          <mat-icon class="mat-icon-rtl-mirror">
-            {{treeControl.isExpanded(node) ? 'expand_more' : 'chevron_right'}}
-          </mat-icon>
-        </button>
-        {{node.filename}} : {{node.type}}
-      </mat-tree-node>
-    </mat-tree>
+    <mat-expansion-panel-header>Flat tree</mat-expansion-panel-header>
```
Comment (crisbeto) [inline src/dev-app/tree/tree-demo.html]: Rather than listing out the examples individually, you can use the `material-example-list` component to render out all of them. This has the advantage that it will include any new example that we add. You can see how it's being used in the `table-demo.ts`.

File: src/lib/tree/data-source/flat-data-source.ts (hunk 2 above)
Comment (crisbeto) [reply src/lib/tree/data-source/flat-data-source.ts]: Would it make sense to allow `ReadonlyArray<T>` here as well? We had a fix a while ago that added it to all the data source usages.

File: src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts [hunk 8]
```diff
@@ -117,33 +53,27 @@ export class FileDatabase {
...
+    return {
+      expandable: !!node.children,
```
Comment (crisbeto) [inline src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts]: Should this have a check for `node.children.length` as well? E.g. if there's a `children` array, but it's empty, we might not want the node to be expandable.

File: src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts [hunk 9]
```diff
@@ -117,33 +53,27 @@ export class FileDatabase {
...
+      name: node.name,
+      level: level,
```
Comment (crisbeto) [inline src/material-examples/cdk-tree-flat/cdk-tree-flat-example.ts]: `{level: level,}` can be reduced to just `{level,}`.

File: src/dev-app/tree/tree-demo-module.ts (hunk 1 above)
Comment (vivian-hu) [reply src/dev-app/tree/tree-demo-module.ts]: This is a different example-module.  The one you talked about is under material-examples.

Comment (jelbourn) [reply src/dev-app/tree/tree-demo-module.ts]: Ah, right, I mixed those up.

File: src/lib/tree/data-source/flat-data-source.ts (hunk 2 above)
Comment (jelbourn) [reply src/lib/tree/data-source/flat-data-source.ts]: That would be good. The fix would be something like "relax allowed types for tree traversal function"
